crates/matrix-sdk-crypto/src/types/events/room_key_withheld.rs

```diff
@@ -0,0 +1,718 @@
+                )
+            }
+            WithheldCode::NoOlm => {
+                RoomKeyWithheldContent::$algorithm(MegolmV1AesSha2WithheldContent::NoOlm(
+                    NoOlmWithheldContent { $sender_key, $from_device, other: Default::default() }
+                        .into(),
+                ))
+            }
+            _ => unreachable!("Can't create an unknown withheld code content"),
+        }
+    };
+}
+
+impl RoomKeyWithheldContent {
+    /// Creates a withheld content from the given info
+    ///
+    /// # Panics
+    ///
+    /// The method will panic if a unsupported algorithm is given. The only
+    /// supported algorithm as of now is `m.megolm.v1.aes-sha2`.
+    pub fn new(
+        algorithm: EventEncryptionAlgorithm,
+        code: WithheldCode,
+        room_id: OwnedRoomId,
+        session_id: String,
+        sender_key: Curve25519PublicKey,
+        from_device: OwnedDeviceId,
+    ) -> Self {
+        let from_device = JsOption::Some(from_device);
+
+        match algorithm {
+            EventEncryptionAlgorithm::MegolmV1AesSha2 => {
+                construct_withheld_content!(
+                    MegolmV1AesSha2,
+                    code,
+                    room_id,
+                    session_id,
+                    sender_key,
+                    from_device
+                )
+            }
+            #[cfg(feature = "experimental-algorithms")]
+            EventEncryptionAlgorithm::MegolmV2AesSha2 => {
+                construct_withheld_content!(
+                    MegolmV2AesSha2,
+                    code,
+                    room_id,
+                    session_id,
+                    sender_key,
+                    from_device
+                )
+            }
+            _ => unreachable!("Unsupported algorithm {algorithm}"),
+        }
+    }
+
+    /// Get the withheld code of this event.
+    pub fn withheld_code(&self) -> WithheldCode {
+        match self {
+            RoomKeyWithheldContent::MegolmV1AesSha2(c) => c.withheld_code(),
+            #[cfg(feature = "experimental-algorithms")]
+            RoomKeyWithheldContent::MegolmV2AesSha2(c) => c.withheld_code(),
+            RoomKeyWithheldContent::Unknown(c) => c.code.to_owned(),
+        }
+    }
+
+    /// Get the algorithm of the room key withheld.
+    pub fn algorithm(&self) -> EventEncryptionAlgorithm {
+        match &self {
+            RoomKeyWithheldContent::MegolmV1AesSha2(_) => EventEncryptionAlgorithm::MegolmV1AesSha2,
+            #[cfg(feature = "experimental-algorithms")]
+            RoomKeyWithheldContent::MegolmV2AesSha2(_) => EventEncryptionAlgorithm::MegolmV2AesSha2,
+            RoomKeyWithheldContent::Unknown(c) => c.algorithm.to_owned(),
+        }
+    }
+}
+
+impl EventType for RoomKeyWithheldContent {
+    const EVENT_TYPE: &'static str = "m.room_key.withheld";
+}
+
+/// A machine-readable code for why the megolm key was not sent.
+#[derive(
+    Eq,
+    Hash,
+    PartialEq,
+    Clone,
```

Comment:
Pretty random order of std derives. I would suggest "core-to-auxiliary" ordering, i.e. more fundamental ones first.

```suggestion
    Clone,
    PartialEq,
    Eq,
    Hash,
```